lang/rust/rust.py

```diff
@@ -1,215 +1,558 @@
+all_traits = {
+    **closure_traits,
+    **conversion_traits,
+    **iterator_traits,
+}
 
-    def code_operator_greater_than():
-        actions.auto(' > ')
 
-    def code_operator_greater_than_or_equal_to():
-        actions.auto(' >= ')
+# tag: libraries_gui
+ctx.lists['user.code_libraries'] = {
+    'eye oh': 'std::io',
+    'file system': 'std::fs',
+    'envy': 'std::env',
+    'collections': 'std::collections',
+}
 
-    def code_operator_less_than():
-        actions.auto(' < ')
+# tag: functions_gui
+ctx.lists['user.code_functions'] = {
+    'drop': 'drop',
+    'catch unwind': 'catch_unwind',
+    'iterator': 'iter',
+    'into iterator': 'into_iter',
+    'from iterator': 'from_iter',
+    **all_macros,
+}
 
-    def code_operator_less_than_or_equal_to():
-        actions.auto(' <= ')
+# tag: functions
+ctx.lists['user.code_type'] = all_types
+
+# rust specific grammar
+ctx.lists['user.code_type_modifier'] = {
+    'mutable': 'mut ',
+    'mute': 'mut ',
+    'borrowed': '&',
+    'borrowed mutable': '&mut ',
+    'borrowed mute': '&mut ',
+    'mutable borrowed': '&mut ',
+    'mute borrowed': '&mut ',
+}
 
-    def code_operator_and():
-        actions.auto(' && ')
 
-    def code_operator_or():
-        actions.auto(' || ')
+@ctx.capture("user.code_type", rule='[{user.code_type_modifier}]{user.code_type}')
+def code_type(m) -> str:
+    """Returns a macro name"""
+    return ''.join(m)
 
-    def code_operator_bitwise_and():
-        actions.auto(' & ')
 
-    def code_operator_bitwise_or():
-        actions.auto(' | ')
+ctx.lists['user.code_macros'] = all_macros
 
-    def code_operator_bitwise_exclusive_or():
-        actions.auto(' ^ ')
+ctx.lists['user.code_trait'] = all_traits
 
-    def code_operator_bitwise_left_shift():
-        actions.auto(' << ')
 
-    def code_operator_bitwise_left_shift_assignment():
-        actions.auto(' <<= ')
+@ctx.action_class('user')
+class UserActions:
 
-    def code_operator_bitwise_right_shift():
-        actions.auto(' >> ')
+    # tag: comment_line
 
-    def code_operator_bitwise_right_shift_assignment():
-        actions.auto(' >>= ')
+    def code_comment_line_prefix():
+        actions.auto_insert('// ')
 
-    def code_operator_object_accessor():
-        actions.auto('.')
+    # tag: comment_documentation
 
-    def code_state_switch():
-        actions.insert('match ')
+    def code_comment_documentation():
+        actions.auto_insert('/// ')
+
+    # tag: imperative
 
     def code_block():
-        actions.auto('{\n\n}')
+        actions.auto_insert('{}')
         actions.edit.left()
+        actions.key('enter')
+
+    def code_state_if():
+        actions.auto_insert('if ')
+
+    def code_state_else_if():
+        actions.auto_insert(' else if ')
+
+    def code_state_else():
+        actions.user.insert_between(' else { ', ' }')
+
+    def code_state_switch():
+        actions.auto_insert('match ')
+
+    def code_state_for():
+        actions.auto_insert('for  in  {}')
```

Comment:
Should the implementations of `code_state_for` and `code_state_for_each` be the same? I don't think the Python implementation does this.

Reply:
I implemented both so that invoking either one would result in the only for statement, however in practice I have found it pointless. I will remove the fall each implementation.